cs50/2017/fall/recover/check50/__init__.py

```diff
@@ -72,25 +49,57 @@ def test_noimage(self):
     @check("compiles")
     def first_image(self):
         """recovers 000.jpg correctly"""
+        self.initialize_data()
         self.add("card.raw")
         self.spawn("./recover card.raw").exit(0, timeout=10)
-        if self.hash("000.jpg") != self.hashes[0]:
-            raise Error("recovered image does not match")
+        self.check_jpg("000.jpg")
 
     @check("compiles")
     def middle_images(self):
         """recovers middle images correctly"""
+        self.initialize_data()
         self.add("card.raw")
         self.spawn("./recover card.raw").exit(0, timeout=10)
-
-        for i, hash in enumerate(self.hashes[1:-1], 1):
-            if hash != self.hash("{:03d}.jpg".format(i)):
-                raise Error("recovered image does not match")
+        for i in range(1, 49):
+            self.check_jpg("{:03d}.jpg".format(i))
 
     @check("compiles")
     def last_image(self):
         """recovers 049.jpg correctly"""
+        self.initialize_data()
         self.add("card.raw")
         self.spawn("./recover card.raw").exit(0, timeout=10)
-        if self.hash("049.jpg") != self.hashes[-1]:
-            raise Error("recovered image does not match")
+        self.check_jpg("049.jpg")
+
+    def check_jpg(self, jpg):
+        """Checks for a jpeg to have the correct hash, and if it does not,
+        raises and error with a helpful hint"""
+        self.require(jpg)
+
+        # pass test if jpeg is correct
+        if self.hash(jpg) == self.hashes[jpg]:
+            return
+
+        err = Error("recovered image does not match")
+
+        # read jpeg header as a byte array and throw out extra bits
+        with open(jpg, "rb") as f:
+            header = bytearray(f.read(4))
+        if len(header) >= 4:
+            header[3] = header[3] & 0xf0
+
+        # test that the jpeg header is correct
+        correct = b'\xff\xd8\xff\xe0'
+        if header != correct:
+            header_str = " ".join(["0x{:02x}".format(b) for b in header])
```

Comment:
You can use generators here for efficiency. `" ".join("0x:{02x}".format(b) for b in header)`